scipy/spatial/transform/_rotation.pyx

```diff
@@ -223,6 +236,124 @@ cdef double[:, :] _compute_euler_from_matrix(
 
     return angles
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef double[:, :] _compute_euler_from_quat(
+    np.ndarray[double, ndim=2] quat, const uchar[:] seq, bint extrinsic=False
+):
+    # The algorithm assumes extrinsic frame transformations. The algorithm
+    # in the paper is formulated for rotation quaternions, which are stored
+    # directly by Rotation.
+    # Adapt the algorithm for our case by
+    # 1. Replacing "d" with "-d" for intrinsic rotation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+    
+    if not extrinsic:
+        seq = seq[::-1]
+
+    cdef int i = _elementary_basis_index(seq[0])
+    cdef int j = _elementary_basis_index(seq[1])
+    cdef int k = _elementary_basis_index(seq[2])
+
+    cdef bint is_proper = i == k
+    if is_proper:
+        k = 3 - i - j # get third axis
+        
+    # Step 0
+    # Check if permutation is even (+1) or odd (-1)     
+    cdef int sign = (i - j) * (j - k) * (k - i) // 2
+
+    cdef Py_ssize_t num_rotations = quat.shape[0]
+
+    # some forward definitions
+    cdef double[:, :] angles = _empty2(num_rotations, 3)
+    cdef double[:] _angles # accessor for each rotation
+    cdef double a, b, c, d
+    cdef double half_sum, half_diff
+    cdef double eps = 1e-7
+    cdef bint safe1, safe2, safe, adjust
+
+    for ind in range(num_rotations):
+        _angles = angles[ind, :]
+
+        # Step 1
+        # Permutate quaternion elements            
+        if is_proper:
+            a = quat[ind, 3]
+            b = quat[ind, i]
+            c = quat[ind, j]
+            d = quat[ind, k] * sign
+        else:
+            a = quat[ind, 3] - quat[ind, j]
+            b = quat[ind, i] + quat[ind, k] * sign
+            c = quat[ind, j] + quat[ind, 3]
+            d = quat[ind, k] * sign - quat[ind, i]
+        
+        # Step 2
+        # Compute second angle...
+        _angles[1] = 2*atan2(sqrt(c*c + d*d), sqrt(a*a + b*b))
+
+        # ... and check if equal to is 0 or pi, causing a singularity
+        safe1 = abs(_angles[1]) >= eps
+        safe2 = abs(_angles[1] - <double>pi) >= eps
+        safe = safe1 and safe2
+
+        # Step 3
+        # compute first and third angles, according to case
+        if safe:
+            half_sum = atan2(b, a) # == (angles[0] + angles[2])/2
+            half_diff = atan2(-d, c) # == (angles[0] - angles[2])/2 
+        
+            _angles[0] = half_sum + half_diff
+            _angles[2] = half_sum - half_diff
+
+        if not extrinsic:
+            # For intrinsic, set first angle to zero so that after reversal we
+            # ensure that third angle is zero
+            # 
+            if not safe:
+                _angles[0] = 0
+            # 
+            if not safe1:
+                _angles[2] = 2 * atan2(b, a)
```

Comment:
This expressions appears above in `safe` branch as well.

We can compute it irrespective of the conditions and reuse.

Reply:
You mean the whole line `_angles[2] = 2 * atan2(b, a)`? If it's the case, then not really, since its either the first or the third angle that gets this value, depending if it's extrinsic or intrinsic rotations.

If you mean the actual calculation of `half_sum` and `half_diff`, then I decided not to always calculate it in order to avoid calling `atan2` at a possibly indefinite case. Don't you think it might create problems?